In [10]:
import pandas as pd

In [11]:
Melb_House_Data=pd.read_csv("C:\Mz-Files\-Mozaik-\Management Tutorial\Machine Learning-Tutorial\melb_data.csv")

In [12]:
Melb_House_Data.head()

Suburb           Address  Rooms Type    Price Method SellerG  \
0  Abbotsford      85 Turner St      2    h  1480000      S  Biggin   
1  Abbotsford   25 Bloomburg St      2    h  1035000      S  Biggin   
2  Abbotsford      5 Charles St      3    h  1465000     SP  Biggin   
3  Abbotsford  40 Federation La      3    h   850000     PI  Biggin   
4  Abbotsford       55a Park St      4    h  1600000     VB  Nelson   

        Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0  3/12/2016       2.5      3067  ...         1  1.0       202           NaN   
1   4/2/2016       2.5      3067  ...         1  0.0       156          79.0   
2   4/3/2017       2.5      3067  ...         2  0.0       134         150.0   
3   4/3/2017       2.5      3067  ...         2  1.0        94           NaN   
4   4/6/2016       2.5      3067  ...         1  2.0       120         142.0   

   YearBuilt  CouncilArea Lattitude  Longtitude             Regionname  \
0        NaN        Yarra  -37.7996    144.9984  Northern Metropolitan   
1     1900.0        Yarra  -37.8079    144.9934  Northern Metropolitan   
2     1900.0        Yarra  -37.8093    144.9944  Northern Metropolitan   
3        NaN        Yarra  -37.7969    144.9969  Northern Metropolitan   
4     2014.0        Yarra  -37.8072    144.9941  Northern Metropolitan   

  Propertycount  
0          4019  
1          4019  
2          4019  
3          4019  
4          4019  

[5 rows x 21 columns]

In [5]:
Melb_House_Data.describe()

Rooms         Price      Distance      Postcode      Bedroom2  \
count  13580.000000  1.358000e+04  13580.000000  13580.000000  13580.000000   
mean       2.937997  1.075684e+06     10.137776   3105.301915      2.914728   
std        0.955748  6.393107e+05      5.868725     90.676964      0.965921   
min        1.000000  8.500000e+04      0.000000   3000.000000      0.000000   
25%        2.000000  6.500000e+05      6.100000   3044.000000      2.000000   
50%        3.000000  9.030000e+05      9.200000   3084.000000      3.000000   
75%        3.000000  1.330000e+06     13.000000   3148.000000      3.000000   
max       10.000000  9.000000e+06     48.100000   3977.000000     20.000000   

           Bathroom           Car       Landsize  BuildingArea    YearBuilt  \
count  13580.000000  13518.000000   13580.000000   7130.000000  8205.000000   
mean       1.534242      1.610075     558.416127    151.967650  1964.684217   
std        0.691712      0.962634    3990.669241    541.014538    37.273762   
min        0.000000      0.000000       0.000000      0.000000  1196.000000   
25%        1.000000      1.000000     177.000000     93.000000  1940.000000   
50%        1.000000      2.000000     440.000000    126.000000  1970.000000   
75%        2.000000      2.000000     651.000000    174.000000  1999.000000   
max        8.000000     10.000000  433014.000000  44515.000000  2018.000000   

          Lattitude    Longtitude  Propertycount  
count  13580.000000  13580.000000   13580.000000  
mean     -37.809203    144.995216    7454.417378  
std        0.079260      0.103916    4378.581772  
min      -38.182550    144.431810     249.000000  
25%      -37.856822    144.929600    4380.000000  
50%      -37.802355    145.000100    6555.000000  
75%      -37.756400    145.058305   10331.000000  
max      -37.408530    145.526350   21650.000000

In [13]:
Melb_House_Data.columns

Index(['Suburb', 'Address', 'Rooms', 'Type', 'Price', 'Method', 'SellerG',
       'Date', 'Distance', 'Postcode', 'Bedroom2', 'Bathroom', 'Car',
       'Landsize', 'BuildingArea', 'YearBuilt', 'CouncilArea', 'Lattitude',
       'Longtitude', 'Regionname', 'Propertycount'],
      dtype='object')

In [14]:
Melb_House_Data=Melb_House_Data.dropna(axis=0)

In [18]:
y=Melb_House_Data.Price

In [ ]:
y.head()

0    1480000
1    1035000
2    1465000
3     850000
4    1600000
Name: Price, dtype: int64

In [14]:
Melb_House_Features=['Rooms', 'Bathroom', 'Landsize', 'Lattitude', 'Longtitude']

In [15]:
X=Melb_House_Data[Melb_House_Features]

In [23]:
X.head()

Rooms  Bathroom  Landsize  Lattitude  Longtitude
1      2         1       156   -37.8079    144.9934
2      3         2       134   -37.8093    144.9944
4      4         1       120   -37.8072    144.9941
6      3         2       245   -37.8024    144.9993
7      2         1       256   -37.8060    144.9954

In [16]:
X.describe()

Rooms      Bathroom       Landsize     Lattitude    Longtitude
count  13580.000000  13580.000000   13580.000000  13580.000000  13580.000000
mean       2.937997      1.534242     558.416127    -37.809203    144.995216
std        0.955748      0.691712    3990.669241      0.079260      0.103916
min        1.000000      0.000000       0.000000    -38.182550    144.431810
25%        2.000000      1.000000     177.000000    -37.856822    144.929600
50%        3.000000      1.000000     440.000000    -37.802355    145.000100
75%        3.000000      2.000000     651.000000    -37.756400    145.058305
max       10.000000      8.000000  433014.000000    -37.408530    145.526350

In [20]:
from sklearn.tree import DecisionTreeRegressor
Melb_House_Model = DecisionTreeRegressor(random_state=1)
Melb_House_Model.fit(X, y)

DecisionTreeRegressor(random_state=1)

In [21]:
print("price prediction for following five houses:")
print(X.head())
print("The predictions are:")
print(Melb_House_Model.predict(X.head())) 


price prediction for following five houses:
   Rooms  Bathroom  Landsize  Lattitude  Longtitude
0      2         1       202   -37.7996    144.9984
1      2         1       156   -37.8079    144.9934
2      3         2       134   -37.8093    144.9944
3      3         2        94   -37.7969    144.9969
4      4         1       120   -37.8072    144.9941
The predictions are:
[1480000. 1035000. 1465000.  850000. 1600000.]


In [30]:
print(y.head())

1    1035000
2    1465000
4    1600000
6    1876000
7    1636000
Name: Price, dtype: int64


In [25]:
from sklearn.metrics import mean_absolute_error

In [26]:
predicted_home_prices = Melb_House_Model.predict(X.head())
mean_absolute_error(y.head(), predicted_home_prices)
MLE_head=mean_absolute_error(y.head(), predicted_home_prices)
print(MLE_head)

0.0


**As we can see, the Actual value is equal with predicted value-MAE=0 (Head)**

In [27]:
predicted_home_prices = Melb_House_Model.predict(X)
mean_absolute_error(y, predicted_home_prices)
MLE=mean_absolute_error(y, predicted_home_prices)
print(MLE)

62000.67361315661


In [37]:
print("price prediction for following five houses:")
print(X)
print("The predictions are:")
print(Melb_House_Model.predict(X)) 


price prediction for following five houses:
       Rooms  Bathroom  Landsize  Lattitude  Longtitude
1          2         1       156  -37.80790   144.99340
2          3         2       134  -37.80930   144.99440
4          4         1       120  -37.80720   144.99410
6          3         2       245  -37.80240   144.99930
7          2         1       256  -37.80600   144.99540
...      ...       ...       ...        ...         ...
12205      3         2       972  -37.51232   145.13282
12206      3         1       179  -37.86558   144.90474
12207      1         1         0  -37.85588   144.89936
12209      2         1         0  -37.85581   144.99025
12212      6         3      1087  -37.81038   144.89389

[6196 rows x 5 columns]
The predictions are:
[1035000. 1465000. 1600000. ...  385000.  560000. 2450000.]


In [38]:
print(y)

1        1035000
2        1465000
4        1600000
6        1876000
7        1636000
          ...   
12205     601000
12206    1050000
12207     385000
12209     560000
12212    2450000
Name: Price, Length: 6196, dtype: int64


In [45]:
print(MLE)

1115.7467183128902


**As we can see, the Actual value is equal with predicted value-MAE=1115 (can be ignored) (Total)**

**Because Model Learned From In-Sample Data**

*Now We want evaluate model with test data instead of training data*

In [28]:
from sklearn.model_selection import train_test_split
train_X, val_X, train_y, val_y = train_test_split(X, y, random_state =0)
Melb_House_Model=DecisionTreeRegressor()
Melb_House_Model.fit(train_X, train_y)
val_predictions = Melb_House_Model.predict(val_X)
print(mean_absolute_error(val_y, val_predictions))


247100.43073146784


**Wow! The error in new data is about a quarter of the average home value.**

**Model is not accurate**

In [59]:
print(val_predictions)  

[ 950000.  696750. 1015000. ...  240000. 1075000. 1055000.]


In [60]:
print(val_y)

4850      815000
2307      655000
10090     957500
3645     1330000
4930      722000
          ...   
8223      520000
11190     870000
8563      200000
1867     1002000
8375     1710000
Name: Price, Length: 1549, dtype: int64


In [61]:
print(train_y)

10385    1060000
5805      390000
8488      502000
6672     1055000
776      1900000
          ...   
9510     1875000
6023      605000
2960      709000
4729     1000000
4996      890000
Name: Price, Length: 4647, dtype: int64


**In fact, y is Union of train_y & val_y**


**and X is Union of train_X & val_X**

*We need to* **Underfitting and Overfitting** *to increase our model accuracy* 

*We define function **get_mae** with **Max_leaf_nodes** for determine size of Decision Tree*

In [1]:
def get_mae(max_leaf_nodes, train_X, val_X, train_y, val_y):
    model = DecisionTreeRegressor(max_leaf_nodes=max_leaf_nodes, random_state=0)
    model.fit(train_X, train_y)
    preds_val = model.predict(val_X)
    mae = mean_absolute_error(val_y, preds_val)
    return(mae)

In [31]:
for max_leaf_nodes in [5, 50, 500, 5000]:
    my_mae = get_mae(max_leaf_nodes, train_X, val_X, train_y, val_y)
    print("Max leaf nodes: %d  \t\t Mean Absolute Error:  %d" %(max_leaf_nodes, my_mae))

Max leaf nodes: 5  		 Mean Absolute Error:  354662
Max leaf nodes: 50  		 Mean Absolute Error:  266447
Max leaf nodes: 500  		 Mean Absolute Error:  231301
Max leaf nodes: 5000  		 Mean Absolute Error:  248846


In [32]:
candidate_max_leaf_nodes = [5, 50, 500, 5000]
scores = {leaf_size: get_mae(leaf_size, train_X, val_X, train_y, val_y) for leaf_size in candidate_max_leaf_nodes}
best_tree_size = min(scores, key=scores.get)
print(best_tree_size)

500


In [34]:
final_model = DecisionTreeRegressor(max_leaf_nodes=best_tree_size, random_state=1)
final_model.fit(train_X, train_y)
pred_val=final_model.predict(val_X)
val_mae = mean_absolute_error(pred_val, val_y)
print("Validation MAE: {:,.0f}".format(val_mae))

Validation MAE: 228,393


**As we can see, the error is decreased. 247,100 to 228,393**


*We also can improve this error with **Random Forests** method* 

In [35]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error

forest_model = RandomForestRegressor(random_state=1)
forest_model.fit(train_X, train_y)
melb_preds = forest_model.predict(val_X)
print(mean_absolute_error(val_y, melb_preds))

180860.37877504269


In [37]:
print(val_y)

8505     2165000
5523      815000
12852     610000
4818     1245000
12812    1160000
          ...   
12961     875000
2572     1286000
9503      885000
356       487500
2034     1194500
Name: Price, Length: 3395, dtype: int64


In [39]:
print(melb_preds)
print(len(melb_preds))

[1673250.  892114.  584750. ... 1461570.  868545. 1013685.]
3395
